### DISCUSSION TITLE
Custom editor for xml files

### DISCUSSION THREAD

**Discussion by HolyFreak2** [eclipse-theia/theia#10646] [Q&A]
Hey everyone,

i am new to Theia and just got it to run locally on my PC. The next thing i would like to do is make my own custom editor for xml files. I would like to convert xml into a graphical view depending on keywords written in the xml file (sth like html). I searched for some input on the internet but didnt find something that helped me with my task. So im trying my luck here, maybe somone knows some website or tutorials that can help me getting started with the task i stated above. Every information will be helpful. Thanks in advance.

**Comment by msujew** [comment]
Hi @HolyFreak2,

so it kinds of depends on what you're trying to accomplish:

* Should the graphical view just be for viewing the XML data, or also for editing?
* Is the graphical view graph based or more complex? (for graph based stuff, I'd recommend [sprotty](https://github.com/eclipse/sprotty))
* Are you targeting only Theia or VSCode as well?

The last question is a bit leading, since you could accomplish your task using a VSCode extension and its [Webview API](https://code.visualstudio.com/api/extension-guides/webview). In general, especially since you mentioned being inexperienced in this regard, I would recommend the custom widget tutorial on the Theia website. Unfortunately, the actual domain seems not to be reachable right now, but you can bypass the domain: https://romantic-mcclintock-8d974f.netlify.app/docs/widgets/

**Reply by HolyFreak2** [reply]
Hey @msujew,

thanks for the quck answer.

So the graphical view should be for editing aswell and it is a bit more complex than just graph based. For your last question i read about using vs code extensions in theia, but i also read that some people have trouble using them in theia. Would it be possible for me to use the API you stated and develop my editor as a vs extension and then add it to my theia instance. 

Also thanks for pointing out the widget tutorial, i will do that first now.
